OpenRA.Mods.Common/Activities/FindResources.cs

```diff
@@ -34,103 +37,112 @@ public FindResources(CPos avoidCell)
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
-
-						// Can the harvester collect this kind of resource?
-						if (!harvInfo.Resources.Contains(resType.Info.Name))
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
-
-						if (territory != null)
-						{
-							// Another harvester has claimed this resource:
-							ResourceClaim claim;
-							if (territory.IsClaimedByAnyoneElse(self, loc, out claim))
-								return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
-						}
-
-						return 0;
-					})
-					.FromPoint(self.Location));
-
-			if (path.Count == 0)
+			// If no harvestable position could be found, then return an empty path
+			if (closestHarvestablePosition == CPos.Zero)
 			{
 				if (!harv.IsEmpty)
 					return new DeliverResources();
+
+				// Get out of the way if we are
+				harv.UnblockRefinery(self);
+				var moveTo = harv.LastHarvestedCell ?? (self.Location + new CVec(0, 4));
+				self.QueueActivity(mobile.MoveTo(moveTo, 1));
+				self.SetTargetLine(Target.FromCell(self.World, moveTo), Color.Gray, false);
+
+				var randFrames = 125 + self.World.SharedRandom.Next(-35, 35);
+				if (NextActivity != null)
+					return Util.SequenceActivities(NextActivity, new Wait(randFrames), new FindResources());
 				else
-				{
-					// Get out of the way if we are:
-					harv.UnblockRefinery(self);
-					var randFrames = 125 + self.World.SharedRandom.Next(-35, 35);
-					if (NextActivity != null)
-						return Util.SequenceActivities(NextActivity, new Wait(randFrames), new FindResources());
-					else
-						return Util.SequenceActivities(new Wait(randFrames), new FindResources());
-				}
+					return Util.SequenceActivities(new Wait(randFrames), new FindResources());
 			}
 
 			// Attempt to claim a resource as ours:
 			if (territory != null)
 			{
-				if (!territory.ClaimResource(self, path[0]))
+				if (!territory.ClaimResource(self, closestHarvestablePosition))
 					return Util.SequenceActivities(new Wait(25), new FindResources());
 			}
 
 			// If not given a direct order, assume ordered to the first resource location we find:
-			if (harv.LastOrderLocation == null)
-				harv.LastOrderLocation = path[0];
+			if (!harv.LastOrderLocation.HasValue)
+				harv.LastOrderLocation = closestHarvestablePosition;
 
-			self.SetTargetLine(Target.FromCell(self.World, path[0]), Color.Red, false);
+			self.SetTargetLine(Target.FromCell(self.World, closestHarvestablePosition), Color.Red, false);
 
 			var notify = self.TraitsImplementing<INotifyHarvesterAction>();
 			var next = new FindResources();
 			foreach (var n in notify)
-				n.MovingToResources(self, path[0], next);
+				n.MovingToResources(self, closestHarvestablePosition, next);
 
-			return Util.SequenceActivities(mobile.MoveTo(path[0], 1), new HarvestResource(), new FindResources());
+			return Util.SequenceActivities(mobile.MoveTo(closestHarvestablePosition, 1), new HarvestResource(), new FindResources());
```

Comment:
The way it is right now was the way it worked previously, but I agree with you. I'll change it and put it under test.